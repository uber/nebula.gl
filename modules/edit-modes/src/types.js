// @flow

import type { Position } from './geojson-types.js';

export type ScreenCoordinates = [number, number];

// Represents an edit action, i.e. a suggestion to update the data based on user interaction events
export type EditAction<TData> = {
  updatedData: TData,
  editType: string,
  editContext: any
};

// Represents an object "picked" from the screen. This usually reflects an object under the cursor
export type Pick = {
  object: any,
  index: number,
  isGuide: boolean
};

// Represents a click event
export type ClickEvent = {
  picks: Pick[],
  screenCoords: ScreenCoordinates,
  mapCoords: Position,
  sourceEvent: any
};

// Represents a double-click event
export type DoubleClickEvent = {
  mapCoords: Position,
  sourceEvent: any
};

// Represents an event that occurs when the pointer goes down and the cursor starts moving
export type StartDraggingEvent = {
  picks: Pick[],
  screenCoords: ScreenCoordinates,
  mapCoords: Position,
  pointerDownScreenCoords: ScreenCoordinates,
  pointerDownMapCoords: Position,
  sourceEvent: any
};

// Represents an event that occurs after the pointer goes down, moves some, then the pointer goes back up
export type StopDraggingEvent = {
  picks: Pick[],
  screenCoords: ScreenCoordinates,
  mapCoords: Position,
  pointerDownScreenCoords: ScreenCoordinates,
  pointerDownMapCoords: Position,
  sourceEvent: any
};

// Represents an event that occurs every time the pointer moves
export type PointerMoveEvent = {
  screenCoords: ScreenCoordinates,
  mapCoords: Position,
  picks: Pick[],
  isDragging: boolean,
  pointerDownPicks: ?(Pick[]),
  pointerDownScreenCoords: ?ScreenCoordinates,
  pointerDownMapCoords: ?Position,
  sourceEvent: any
};
