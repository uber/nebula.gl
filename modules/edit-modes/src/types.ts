import { Position, Point, Geometry, FeatureWithProps } from './geojson-types';

export type ScreenCoordinates = [number, number];

// Represents an edit action, i.e. a suggestion to update the data based on user interaction events
export type EditAction<TData> = {
  updatedData: TData;
  editType: string;
  editContext: any;
};

// Represents an object "picked" from the screen. This usually reflects an object under the cursor
export type Pick = {
  index: number;
  object?: any;
  isGuide?: boolean;
  featureIndex?: number;
  type?: string;
};

export type Viewport = {
  width: number;
  height: number;
  longitude: number;
  latitude: number;
  zoom: number;
  bearing?: number;
  pitch?: number;
};

export type BasePointerEvent = {
  picks: Pick[];
  screenCoords: ScreenCoordinates;
  mapCoords: Position;
  sourceEvent: any;
};

// Represents a click event
export type ClickEvent = BasePointerEvent;

// Represents an event that occurs when the pointer goes down and the cursor starts moving
export type StartDraggingEvent = BasePointerEvent & {
  pointerDownPicks?: Pick[] | null | undefined;
  pointerDownScreenCoords: ScreenCoordinates;
  pointerDownMapCoords: Position;
  cancelPan: () => void;
};

// Represents an event that occurs after the pointer goes down, moves some, then the pointer goes back up
export type StopDraggingEvent = BasePointerEvent & {
  pointerDownPicks?: Pick[] | null | undefined;
  pointerDownScreenCoords: ScreenCoordinates;
  pointerDownMapCoords: Position;
};

// Represents an event that occurs after the pointer goes down and is moving
export type DraggingEvent = BasePointerEvent & {
  pointerDownPicks?: Pick[] | null | undefined;
  pointerDownScreenCoords: ScreenCoordinates;
  pointerDownMapCoords: Position;
  cancelPan: () => void;
};

// Represents an event that occurs every time the pointer moves
export type PointerMoveEvent = BasePointerEvent & {
  pointerDownPicks?: Pick[] | null | undefined;
  pointerDownScreenCoords?: ScreenCoordinates | null | undefined;
  pointerDownMapCoords?: Position | null | undefined;
  cancelPan: () => void;
};

export type Tooltip = {
  position: Position;
  text: string;
};

export type EditHandleType =
  | 'existing'
  | 'intermediate'
  | 'snap-source'
  | 'snap-target'
  | 'scale'
  | 'rotate';

export type EditHandleFeature = FeatureWithProps<
  Point,
  {
    guideType: 'editHandle';
    editHandleType: EditHandleType;
    featureIndex: number;
    positionIndexes?: number[];
    shape?: string;
  }
>;

export type TentativeFeature = FeatureWithProps<
  Geometry,
  {
    guideType: 'tentative';
    shape?: string;
  }
>;

export type GuideFeature = EditHandleFeature | TentativeFeature;

export type GuideFeatureCollection = {
  type: 'FeatureCollection';
  features: Readonly<GuideFeature>[];
  properties?: {};
};

export type ModeProps<TData> = {
  // The data being edited, this can be an array or an object
  data: TData;

  // Additional configuration for this mode
  modeConfig: any;

  // The indexes of the selected features
  selectedIndexes: number[];

  // The cursor type, as a [CSS Cursor](https://developer.mozilla.org/en-US/docs/Web/CSS/cursor)
  cursor: string | null | undefined;

  // The last pointer move event that occurred
  lastPointerMoveEvent: PointerMoveEvent;

  // Callback used to notify applications of an edit action
  onEdit: (editAction: EditAction<TData>) => void;

  // Callback used to update cursor
  onUpdateCursor: (cursor: string | null | undefined) => void;
};
