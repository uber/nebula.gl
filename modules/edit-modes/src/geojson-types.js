// @flow

// Geometry types

export type Position = [number, number] | [number, number, number];
export type PointCoordinates = Position;
export type LineStringCoordinates = Position[];
export type PolygonCoordinates = Position[][];
export type MultiPointCoordinates = Position[];
export type MultiLineStringCoordinates = Position[][];
export type MultiPolygonCoordinates = Position[][][];

export type AnyCoordinates =
  | PointCoordinates
  | LineStringCoordinates
  | PolygonCoordinates
  | MultiPointCoordinates
  | MultiLineStringCoordinates
  | MultiPolygonCoordinates;

export type Point = {
  type: 'Point',
  coordinates: PointCoordinates
};

export type LineString = {
  type: 'LineString',
  coordinates: LineStringCoordinates
};

export type Polygon = {
  type: 'Polygon',
  coordinates: PolygonCoordinates
};

export type MultiPoint = {
  type: 'MultiPoint',
  coordinates: MultiPointCoordinates
};

export type MultiLineString = {
  type: 'MultiLineString',
  coordinates: MultiLineStringCoordinates
};

export type MultiPolygon = {
  type: 'MultiPolygon',
  coordinates: MultiPolygonCoordinates
};

export type Geometry = Point | LineString | Polygon | MultiPoint | MultiLineString | MultiPolygon;

export type Polygonal = Polygon | MultiPolygon;

// Feature types

export type BoundingBoxArray = [number, number, number, number];

export type FeatureOf<T: Geometry> = {
  type: 'Feature',
  geometry: T,
  properties?: {},
  id?: string | number,
  bbox?: BoundingBoxArray
};

export type FeatureWithProps<T: Geometry, P> = {
  type: 'Feature',
  geometry: T,
  properties: P
};

export type Feature =
  | FeatureOf<Point>
  | FeatureOf<LineString>
  | FeatureOf<Polygon>
  | FeatureOf<MultiPoint>
  | FeatureOf<MultiLineString>
  | FeatureOf<MultiPolygon>;

export type FeatureCollection = {
  type: 'FeatureCollection',
  features: Feature[],
  properties?: {},
  id?: string | number,
  bbox?: BoundingBoxArray
};

export type AnyGeoJson = Feature | FeatureCollection;
