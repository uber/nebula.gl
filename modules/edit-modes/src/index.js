// @flow

export { EditMode, BaseEditMode } from './lib/edit-mode.js';
export { ViewMode } from './lib/view-mode.js';
export { DrawPolygonMode } from './lib/draw-polygon-mode.js';
export { ImmutableFeatureCollection } from './lib/immutable-feature-collection.js';

export type { ModeState } from './lib/edit-mode.js';

export type {
  ScreenCoordinates,
  EditAction,
  Pick,
  ClickEvent,
  PointerMoveEvent,
  StartDraggingEvent,
  StopDraggingEvent
} from './types.js';

export type {
  Position,
  PointCoordinates,
  LineStringCoordinates,
  PolygonCoordinates,
  MultiPointCoordinates,
  MultiLineStringCoordinates,
  MultiPolygonCoordinates,
  AnyCoordinates,
  Point,
  LineString,
  Polygon,
  MultiPoint,
  MultiLineString,
  MultiPolygon,
  Geometry,
  Polygonal,
  BoundingBoxArray,
  FeatureOf,
  FeatureWithProps,
  Feature,
  FeatureCollection,
  AnyGeoJson
} from './geojson-types.js';
