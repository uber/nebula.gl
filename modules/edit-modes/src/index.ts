export type { EditMode } from './lib/edit-mode';
export type { GeoJsonEditModeType } from './lib/geojson-edit-mode';

export { GeoJsonEditMode } from './lib/geojson-edit-mode';

// Alter modes
export { ModifyMode } from './lib/modify-mode';
export { TranslateMode } from './lib/translate-mode';
export { ScaleMode } from './lib/scale-mode';
export { RotateMode } from './lib/rotate-mode';
export { DuplicateMode } from './lib/duplicate-mode';
export { ExtendLineStringMode } from './lib/extend-line-string-mode';
export { SplitPolygonMode } from './lib/split-polygon-mode';
export { ExtrudeMode } from './lib/extrude-mode';
export { ElevationMode } from './lib/elevation-mode';
export { TransformMode } from './lib/transform-mode';

// Draw modes
export { DrawPointMode } from './lib/draw-point-mode';
export { DrawLineStringMode } from './lib/draw-line-string-mode';
export { DrawPolygonMode } from './lib/draw-polygon-mode';
export { DrawRectangleMode } from './lib/draw-rectangle-mode';
export { DrawCircleByDiameterMode } from './lib/draw-circle-by-diameter-mode';
export { DrawCircleFromCenterMode } from './lib/draw-circle-from-center-mode';
export { DrawEllipseByBoundingBoxMode } from './lib/draw-ellipse-by-bounding-box-mode';
export { DrawEllipseUsingThreePointsMode } from './lib/draw-ellipse-using-three-points-mode';
export { DrawRectangleUsingThreePointsMode } from './lib/draw-rectangle-using-three-points-mode';
export { Draw90DegreePolygonMode } from './lib/draw-90degree-polygon-mode';
export { DrawPolygonByDraggingMode } from './lib/draw-polygon-by-dragging-mode';
export { ImmutableFeatureCollection } from './lib/immutable-feature-collection';

// Other modes
export { ViewMode } from './lib/view-mode';
export { MeasureDistanceMode } from './lib/measure-distance-mode';
export { MeasureAreaMode } from './lib/measure-area-mode';
export { MeasureAngleMode } from './lib/measure-angle-mode';
export { CompositeMode } from './lib/composite-mode';
export { SnappableMode } from './lib/snappable-mode';

// Experimental
export {default as _memoize} from './memoize';

export type {
  ScreenCoordinates,
  EditAction,
  Pick,
  ClickEvent,
  PointerMoveEvent,
  StartDraggingEvent,
  StopDraggingEvent,
  DraggingEvent,
  DoubleClickEvent,
  ModeProps,
  GuideFeatureCollection,
  Viewport,
  Tooltip,
} from './types';

export type {
  Position,
  PointCoordinates,
  LineStringCoordinates,
  PolygonCoordinates,
  MultiPointCoordinates,
  MultiLineStringCoordinates,
  MultiPolygonCoordinates,
  AnyCoordinates,
  Point,
  LineString,
  Polygon,
  MultiPoint,
  MultiLineString,
  MultiPolygon,
  Geometry,
  Polygonal,
  BoundingBoxArray,
  FeatureOf,
  FeatureWithProps,
  Feature,
  FeatureCollection,
  AnyGeoJson,
} from './geojson-types';
