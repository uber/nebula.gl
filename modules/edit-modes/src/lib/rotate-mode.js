// @flow

import turfCentroid from '@turf/centroid';
import turfBearing from '@turf/bearing';
import turfTransformRotate from '@turf/transform-rotate';
import type {
  PointerMoveEvent,
  StartDraggingEvent,
  StopDraggingEvent,
  DraggingEvent,
  ModeProps
} from '../types.js';
import type { FeatureCollection, Position } from '../geojson-types.js';
import { BaseGeoJsonEditMode, type GeoJsonEditAction } from './geojson-edit-mode.js';
import { ImmutableFeatureCollection } from './immutable-feature-collection.js';

export class RotateMode extends BaseGeoJsonEditMode {
  _isRotatable: boolean;
  _geometryBeingRotated: ?FeatureCollection;

  handleDragging(event: DraggingEvent, props: ModeProps<FeatureCollection>) {
    if (!this._isRotatable) {
      // Nothing to do
      return;
    }

    if (this._geometryBeingRotated) {
      // Rotate the geometry
      props.onEdit(
        this.getRotateAction(event.pointerDownMapCoords, event.mapCoords, 'rotating', props)
      );
    }

    // cancel map panning
    event.cancelPan();
  }

  handlePointerMove(event: PointerMoveEvent, props: ModeProps<FeatureCollection>) {
    this._isRotatable = this.isSelectionPicked(event.pointerDownPicks || event.picks, props);

    this.updateCursor(props);
  }

  handleStartDragging(event: StartDraggingEvent, props: ModeProps<FeatureCollection>) {
    if (!this._isRotatable) {
      return;
    }

    this._geometryBeingRotated = this.getSelectedFeaturesAsFeatureCollection(props);
  }

  handleStopDragging(event: StopDraggingEvent, props: ModeProps<FeatureCollection>) {
    if (this._geometryBeingRotated) {
      // Rotate the geometry
      props.onEdit(
        this.getRotateAction(event.pointerDownMapCoords, event.mapCoords, 'rotated', props)
      );
      this._geometryBeingRotated = null;
    }
  }

  updateCursor(props: ModeProps<FeatureCollection>) {
    if (this._isRotatable) {
      // TODO: look at doing SVG cursors to get a better "rotate" cursor
      props.onUpdateCursor('move');
    } else {
      props.onUpdateCursor(null);
    }
  }

  getRotateAction(
    startDragPoint: Position,
    currentPoint: Position,
    editType: string,
    props: ModeProps<FeatureCollection>
  ): GeoJsonEditAction {
    const startPosition = startDragPoint;
    const centroid = turfCentroid(this._geometryBeingRotated);
    const angle = getRotationAngle(centroid, startPosition, currentPoint);

    const rotatedFeatures = turfTransformRotate(this._geometryBeingRotated, angle);

    let updatedData = new ImmutableFeatureCollection(props.data);

    const selectedIndexes = props.selectedIndexes;
    for (let i = 0; i < selectedIndexes.length; i++) {
      const selectedIndex = selectedIndexes[i];
      const movedFeature = rotatedFeatures.features[i];
      updatedData = updatedData.replaceGeometry(selectedIndex, movedFeature.geometry);
    }

    return {
      updatedData: updatedData.getObject(),
      editType,
      editContext: {
        featureIndexes: selectedIndexes
      }
    };
  }
}

function getRotationAngle(centroid: Position, startDragPoint: Position, currentPoint: Position) {
  const bearing1 = turfBearing(centroid, startDragPoint);
  const bearing2 = turfBearing(centroid, currentPoint);
  return bearing2 - bearing1;
}
