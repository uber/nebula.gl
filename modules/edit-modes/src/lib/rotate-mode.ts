/* eslint-disable prettier/prettier */
import bbox from '@turf/bbox';
import turfCentroid from '@turf/centroid';
import turfBearing from '@turf/bearing';
import bboxPolygon from '@turf/bbox-polygon';
import turfDistance from '@turf/distance';
import { coordEach } from '@turf/meta';
import { getGeom } from '@turf/invariant';
import { point, featureCollection, lineString } from '@turf/helpers';
import turfTransformRotate from '@turf/transform-rotate';
import polygonToLine from '@turf/polygon-to-line';
import {
  PointerMoveEvent,
  StartDraggingEvent,
  StopDraggingEvent,
  DraggingEvent,
  ModeProps,
  EditHandleFeature,
  GuideFeatureCollection,
} from '../types';
import { getPickedEditHandle } from '../utils';
import { FeatureCollection, Position } from '../geojson-types';
import { GeoJsonEditMode, GeoJsonEditAction, getIntermediatePosition } from './geojson-edit-mode';
import { ImmutableFeatureCollection } from './immutable-feature-collection';

export class RotateMode extends GeoJsonEditMode {
  _selectedEditHandle: EditHandleFeature | null | undefined;
  _geometryBeingRotated: FeatureCollection | null | undefined;
  _isRotating = false;

  _isSinglePointGeometrySelected = (geometry: FeatureCollection | null | undefined): boolean => {
    const { features } = geometry || {};
    if (Array.isArray(features) && features.length === 1) {
      // @ts-expect-error turf type diff
      const { type }: { type: string } = getGeom(features[0]);
      return type === 'Point';
    }
    return false;
  };

  getIsRotating = () => this._isRotating;

  getGuides(props: ModeProps<FeatureCollection>): GuideFeatureCollection {
    const selectedGeometry =
      this._geometryBeingRotated || this.getSelectedFeaturesAsFeatureCollection(props);

    if (this._isSinglePointGeometrySelected(selectedGeometry)) {
      return { type: 'FeatureCollection', features: [] };
    }

    if (this._isRotating) {
      // Display rotate pivot
      return featureCollection([turfCentroid(selectedGeometry)]) as GuideFeatureCollection;
    }

    const boundingBox = bboxPolygon(bbox(selectedGeometry));

    let previousCoord = null;
    let topEdgeMidpointCoords = null;
    let longestEdgeLength = 0;

    coordEach(boundingBox, (coord: Position) => {
      if (previousCoord) {
        const edgeMidpoint = getIntermediatePosition(coord, previousCoord);
        if (!topEdgeMidpointCoords || edgeMidpoint[1] > topEdgeMidpointCoords[1]) {
          // Get the top edge midpoint of the enveloping box
          topEdgeMidpointCoords = edgeMidpoint;
        }
        // Get the length of the longest edge of the enveloping box
        const edgeDistance = turfDistance(coord, previousCoord);
        longestEdgeLength = Math.max(longestEdgeLength, edgeDistance);
      }
      previousCoord = coord;
    });

    // Scale the length of the line between the rotate handler and the enveloping box
    // relative to the length of the longest edge of the enveloping box
    const rotateHandleCoords = topEdgeMidpointCoords && [
      topEdgeMidpointCoords[0],
      topEdgeMidpointCoords[1] + longestEdgeLength / 1000,
    ];

    const lineFromEnvelopeToRotateHandle = lineString([topEdgeMidpointCoords, rotateHandleCoords]);
    const rotateHandle = point(rotateHandleCoords, {
      guideType: 'editHandle',
      editHandleType: 'rotate',
    });

    const outFeatures = [
      polygonToLine(boundingBox),
      rotateHandle,
      lineFromEnvelopeToRotateHandle,
    ];

    // @ts-expect-error turf type diff
    return featureCollection(outFeatures);
  }

  handleDragging(event: DraggingEvent, props: ModeProps<FeatureCollection>) {
    if (!this._isRotating) {
      return;
    }

    const rotateAction = this.getRotateAction(
      event.pointerDownMapCoords,
      event.mapCoords,
      'rotating',
      props
    );
    if (rotateAction) {
      props.onEdit(rotateAction);
    }

    event.cancelPan();
  }

  handlePointerMove(event: PointerMoveEvent, props: ModeProps<FeatureCollection>) {
    if (!this._isRotating) {
      const selectedEditHandle = getPickedEditHandle(event.picks);
      this._selectedEditHandle =
        selectedEditHandle && selectedEditHandle.properties.editHandleType === 'rotate'
          ? selectedEditHandle
          : null;
    }

    this.updateCursor(props);
  }

  handleStartDragging(event: StartDraggingEvent, props: ModeProps<FeatureCollection>) {
    if (this._selectedEditHandle) {
      this._isRotating = true;
      this._geometryBeingRotated = this.getSelectedFeaturesAsFeatureCollection(props);
    }
  }

  handleStopDragging(event: StopDraggingEvent, props: ModeProps<FeatureCollection>) {
    if (this._isRotating) {
      // Rotate the geometry
      const rotateAction = this.getRotateAction(
        event.pointerDownMapCoords,
        event.mapCoords,
        'rotated',
        props
      );

      if (rotateAction) {
        props.onEdit(rotateAction);
      }

      this._geometryBeingRotated = null;
      this._selectedEditHandle = null;
      this._isRotating = false;
    }
  }

  updateCursor(props: ModeProps<FeatureCollection>) {
    if (this._selectedEditHandle) {
      // TODO: look at doing SVG cursors to get a better "rotate" cursor
      props.onUpdateCursor('crosshair');
    } else {
      props.onUpdateCursor(null);
    }
  }

  getRotateAction(
    startDragPoint: Position,
    currentPoint: Position,
    editType: string,
    props: ModeProps<FeatureCollection>
  ): GeoJsonEditAction | null | undefined {
    if (!this._geometryBeingRotated) {
      return null;
    }

    const centroid = turfCentroid(this._geometryBeingRotated);
    const angle = getRotationAngle(centroid, startDragPoint, currentPoint);

    // @ts-expect-error turf types too wide
    const rotatedFeatures: FeatureCollection = turfTransformRotate(
      // @ts-expect-error turf types too wide
      this._geometryBeingRotated,
      angle,
      {
        pivot: centroid,
      }
    );

    let updatedData = new ImmutableFeatureCollection(props.data);

    const selectedIndexes = props.selectedIndexes;
    for (let i = 0; i < selectedIndexes.length; i++) {
      const selectedIndex = selectedIndexes[i];
      const movedFeature = rotatedFeatures.features[i];
      updatedData = updatedData.replaceGeometry(selectedIndex, movedFeature.geometry);
    }

    return {
      updatedData: updatedData.getObject(),
      editType,
      editContext: {
        featureIndexes: selectedIndexes,
      },
    };
  }
}

function getRotationAngle(centroid: Position, startDragPoint: Position, currentPoint: Position) {
  const bearing1 = turfBearing(centroid, startDragPoint);
  const bearing2 = turfBearing(centroid, currentPoint);
  return bearing2 - bearing1;
}
