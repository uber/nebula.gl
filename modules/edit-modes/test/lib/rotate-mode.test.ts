import { RotateMode } from '../../src/lib/rotate-mode';
import {
  createFeatureCollectionProps,
  createPointerMoveEvent,
  createStartDraggingEvent,
  createStopDraggingEvent,
} from '../test-utils';
import { Pick, ModeProps } from '../../src/types';
import { FeatureCollection } from '../../src/geojson-types';

let rotateMode: RotateMode;

let warnBefore;
beforeEach(() => {
  warnBefore = console.warn; // eslint-disable-line
  // $FlowFixMe
  console.warn = function () {}; // eslint-disable-line
  rotateMode = new RotateMode();
});

afterEach(() => {
  // $FlowFixMe
  console.warn = warnBefore; // eslint-disable-line
});

const mockRotate = (picks: Pick[], props: ModeProps<FeatureCollection>) => {
  rotateMode.getGuides(props);

  const moveEvent = createPointerMoveEvent([-1, -1], picks);
  rotateMode.handlePointerMove(moveEvent, props);

  const startDragEvent = createStartDraggingEvent([-1, -1], [-1, -1], picks);
  rotateMode.handleStartDragging(startDragEvent, props);

  const stopDragEvent = createStopDraggingEvent([2, 3], [-1, -1], picks);
  rotateMode.handleStopDragging(stopDragEvent, props);
};

test('Selected polygon feature can be rotated', () => {
  const mockOnEdit = jest.fn();
  const props = createFeatureCollectionProps({
    selectedIndexes: [2],
    onEdit: mockOnEdit,
  });
  const mockPicks = [
    {
      index: 2,
      isGuide: true,
      object: {
        type: 'Feature',
        geometry: { type: 'Point', coordinates: [-2, -2] },
        properties: { guideType: 'editHandle', editHandleType: 'rotate' },
      },
    },
  ];
  mockRotate(mockPicks, props);

  expect(mockOnEdit).toHaveBeenCalledTimes(1);
  const scaledFeature = mockOnEdit.mock.calls[0][0].updatedData.features[2];
  expect(scaledFeature).toMatchSnapshot();
  expect(props.data.features[2]).not.toEqual(scaledFeature);
});

test('Selected polygon feature without edit handle picks cannot be rotated', () => {
  const mockOnEdit = jest.fn();
  const props = createFeatureCollectionProps({
    selectedIndexes: [2],
    onEdit: mockOnEdit,
  });

  mockRotate([], props);
  expect(mockOnEdit).toHaveBeenCalledTimes(0);
});
