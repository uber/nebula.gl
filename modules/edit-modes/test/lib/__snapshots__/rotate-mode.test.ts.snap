// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selected polygon feature can be rotated 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        Array [
          0.7839526552529605,
          1.177031499163895,
        ],
        Array [
          -1.1770284241941908,
          0.7839572723738235,
        ],
        Array [
          -0.7839526552529605,
          -1.177031499163893,
        ],
        Array [
          1.1770284241941908,
          -0.783957272373826,
        ],
        Array [
          0.7839526552529605,
          1.177031499163895,
        ],
      ],
      Array [
        Array [
          0.3919743169615231,
          0.5885194916799502,
        ],
        Array [
          0.5885191073002716,
          -0.39197489408891184,
        ],
        Array [
          -0.3919743169615231,
          -0.5885194916799487,
        ],
        Array [
          -0.5885191073002716,
          0.39197489408890945,
        ],
        Array [
          0.3919743169615231,
          0.5885194916799502,
        ],
      ],
    ],
    "type": "Polygon",
  },
  "properties": Object {},
  "type": "Feature",
}
`;
